R/intervals.R

```diff
@@ -0,0 +1,625 @@
+#' @title Generate columns to help describe numeric variable intervals
+#'
+#' @description demUtils assumes numeric interval variables are grouped into
+#' left-closed, right-open intervals. \eqn{a <= x < b}. Each interval can be
+#' described by their endpoints from which interval lengths and nicer formatted
+#' interval names can be created.
+#'
+#' @param dt \[`data.table()`\]\cr
+#'   `col_stem`-specific data.
+#' @param id_cols \[`character()`\]\cr
+#'   ID columns that uniquely identify each row of `dt`. This must include
+#'   `{col_stem}_start`.
+#' @param col_stem \[`character(1)`\]\cr
+#'   Base name of the numeric variable column. Does not include the '_start',
+#'   '_end' etc. suffix.
+#' @param right_most_endpoint \[`numeric(1)`\]\cr
+#'   Assumed right most endpoint of `{col_stem}_end`. Default is \code{Inf}.
+#' @param format \[`character(1)`\]\cr
+#'   Formatting style for the interval names. Default is 'to'; can also be
+#'   'interval' or 'dash'.
+#' @param format_infinite \[`character(1)`\]\cr
+#'   Formatting style for infinite endpoint intervals. Default is 'plus'; can
+#'   also be '+'. Ignored when `format = "interval"`.
+#'
+#' @return Invisibly returns reference to modified `dt`.
+#'
+#' @details `gen_end` generates a new column `{col_stem}_end` for the
+#' right-open endpoint of each interval from a series of left-closed endpoints
+#' `{col_stem}_start`.
+#'
+#' `gen_end` assumes that only the most detailed intervals are present in the
+#' input dataset; including overlapping intervals will not return expected
+#' results.
+#'
+#' Input data `dt` for `gen_end` must:
+#'   * Contain all columns specified in `id_cols`.
+#'   * Have a column called `{col_stem}_start`.
+#'   * Have each row uniquely identified by each combination of `id_cols`.
+#'
+#' `gen_length` generates a new column `{col_stem}_length` for the length of
+#' each interval. Input data `dt` for `gen_length` must contain
+#' `{col_stem}_start` and `{col_stem}_end` columns.
+#'
+#' `gen_name` generates a new column `{col_stem}_name` describing each interval.
+#'
+#' Formatting style for intervals:
+#'   * \eqn{[a, b)} interval notation is used when `format = 'interval`.
+#'   * `a to b` is used when `format = "to"`.
+#'   * `a-b` is used when `format = "dash"`.
```

Comment:
Should we consider the edge case where an interval might include a negative value (and end up with an interval like `"-10--5"`, or is that too far outside our scope?

Reply:
Maybe for now outside our scope but good thing to consider in the future.